Rubberduck.Core/CodeAnalysis/CodeMetrics/CodeMetricsViewModel.cs

```diff
@@ -0,0 +1,186 @@
+                Projects = new ObservableCollection<CodeExplorerItemViewModel>();
+            }
+
+            IsBusy = _state.Status != ParserState.Pending && _state.Status <= ParserState.ResolvedDeclarations;
+
+            var userDeclarations = _state.DeclarationFinder.AllUserDeclarations
+                .GroupBy(declaration => declaration.ProjectId)
+                .ToList();
+
+            if (userDeclarations.Any(
+                    grouping => grouping.All(declaration => declaration.DeclarationType != DeclarationType.Project)))
+            {
+                return;
+            }
+
+            var newProjects = userDeclarations.Select(grouping =>
+                new CodeExplorerProjectViewModel(_folderHelper,
+                    grouping.SingleOrDefault(declaration => declaration.DeclarationType == DeclarationType.Project),
+                    grouping)).ToList();
+
+            UpdateNodes(Projects, newProjects);
+
+            Projects = new ObservableCollection<CodeExplorerItemViewModel>(newProjects);
+
+            IsBusy = false;
+        }
+
+        private void UpdateNodes(IEnumerable<CodeExplorerItemViewModel> oldList, IEnumerable<CodeExplorerItemViewModel> newList)
+        {
+            foreach (var item in newList)
+            {
+                CodeExplorerItemViewModel oldItem;
+
+                if (item is CodeExplorerCustomFolderViewModel)
+                {
+                    oldItem = oldList.FirstOrDefault(i => i.Name == item.Name);
+                }
+                else
+                {
+                    oldItem = oldList.FirstOrDefault(i =>
+                        item.QualifiedSelection != null && i.QualifiedSelection != null &&
+                        i.QualifiedSelection.Value.QualifiedName.ProjectId ==
+                        item.QualifiedSelection.Value.QualifiedName.ProjectId &&
+                        i.QualifiedSelection.Value.QualifiedName.ComponentName ==
+                        item.QualifiedSelection.Value.QualifiedName.ComponentName &&
+                        i.QualifiedSelection.Value.Selection == item.QualifiedSelection.Value.Selection);
+                }
+
+                if (oldItem != null)
+                {
+                    item.IsExpanded = oldItem.IsExpanded;
+                    item.IsSelected = oldItem.IsSelected;
+
+                    if (oldItem.Items.Any() && item.Items.Any())
+                    {
+                        UpdateNodes(oldItem.Items, item.Items);
+                    }
+                }
+            }
+        }
+
+        public void Dispose()
+        {
+            _state.StateChanged -= OnStateChanged;
+        }
+
+        private Dictionary<Declaration, List<ICodeMetricResult>> resultsByDeclaration;
+
+        private CodeExplorerItemViewModel _selectedItem;
+        public CodeExplorerItemViewModel SelectedItem
+        {
+            get => _selectedItem;
+            set
+            {
+                _selectedItem = value;
```

Comment:
if `value == _selectedItem`, we're firing extraneous `PropertyChanged` events here.